docs/server_notifications.md

```diff
@@ -0,0 +1,124 @@
+# Server notifications
+
+Client SDKs communicate with Jellyfish using so called Media Events (ME), which are messages exchanged using Websockets.
+
+A few examples when Media events are sent:
+* `join` - sent when peer joins WebRTC room
+* `peerDenied` - sent if peer was rejected by server when joining to server
+* `tracksAdded` - sent when some tracks (video or audio) were added by a peer
+
+Those messages have to be serialized in some way.
+
+## Data serialization
+
+In general we have two major options here:
+
+### JSON
+
+JSON has been used for a long time in [membrane_rtc_engine](https://github.com/jellyfish-dev/membrane_rtc_engine).
+
+This solution has several drawbacks:
+* there is no versioning so it’s hard to track which version of ME is used on the server side and which one on the client side
+* each client SDK has to implement type definitions, serialization and deserialization on its own. Adding new media events requires a lot of the same work in each client library
+* introducing changes in already existing ME is error-prone
+
+This renders JSON on it's own unsuitable for ME.
+
+However, there are ways to harness JSON and make it easier to manage - both when it comes to versioning and generating code.
+
+#### GraphQL schema
+
+The GraphQL specification defines a human-readable schema definition language (or SDL) that you use to define your schema and store it as a string.
+
+
+```graphql
+type Book {
+  title: String
+  author: Author
+}
+
+type Author {
+  name: String
+  books: [Book]
+}
+```
+
+A schema defines a collection of types and the relationships between those types.
+
+Using the schema code can be generated, for example using the [GraphQL Code Generator].
+In Elixir, using [Absinthe](https://hexdocs.pm/absinthe/overview.html) (a GraphQL toolkit for Elixir) schema is created as an Elixir module, from which Graphql schema can be generated - not the other way around.
+That means, that we would first create schema in Elixir and then generate types for each client SDK.
+
+Importantly, GraphQL allows for deprecating fields. When greater changes are introduced it might be easier to introduce new endpoint with newer API.
+
+Features:
+- generating type specification from schema
+- backward compatibility (deprecating fields)
```

Comment:
> I don't think the topic of GraphQL is sufficiently covered here. GraphQL schemas do look promising on their own, however, there is an issue: we need bidirectional communication and as far as I know, GraphQL requests are usually sent over a simple HTTP request - this only takes care of one direction. For Server -> Client notifications, we would need to use the subscriptions mechanism.

Yep, we'd have to use subscriptions. However, the most painful thing might be synchronization. I really miss the possibility to return data on a successful subscription. In another project, we had trouble since there's a race condition between getting a state and subscribing to updates. E.g., in case of simultaneous joining of 2 peers one could get a state without the new user but omit the event about new peer joining.

> 1. We would have MEs split over two channels.
> 2. HTTP requests will be slower than WebSockets, so the potential for race conditions is increased
> 
> Maybe it would be possible to use a single WebSocket for both subscriptions and queries, but this needs verification.

This is possible, it is even the default behaviour and requires extra configuration to change (https://www.apollographql.com/docs/react/data/subscriptions/#3-split-communication-by-operation-recommended). This IMO doesn't change anything when it comes to RCs

> On the server, we also monitor the signaling to make sure that the peer is still alive. I'm not sure if this is something that GraphQL can provide us with.

It can and can't at the same time :stuck_out_tongue: You have a WebSocket that you can monitor and we were making use of that, but the problem was that Absinthe at the time (dunno how is it now) didn't expose socket pid, so we had to hack the `Absinthe.Phoenix.Socket` to allow that

Reply:
> Can't the state be sent as a first event over the subscription?

This might be more of a problem with Elixir implementation (Absinthe), but the issue is that on subscribe call inside a lambda, you just decide which topic you subscribe to and you cannot return any value. You could contact the process having the state from that lambda, but once again that could be a race if that process sent the response with a state before the subscription is completed, plus that event would be received by anyone subscribing to that topic, which in case of a room-wide topic would mean every participant.

Bear in mind it's been some time since that project, so some things might have changed or I've misremembered something, but I do remember we were getting a feeling we were misusing the GraphQL a bit and struggling with things that wouldn't be a problem without that technology.